### DISCUSSION TITLE
Proposal: Implement file scoped classes

### DISCUSSION THREAD

**Discussion by jguadagno** [dotnet/csharplang#5633] [Language Ideas]
I want to take the file scoped namespaces (#137 ) one step further based on this [tweet](https://twitter.com/haacked/status/1480927254876205058) and have file scoped namespace and classes.

Now with c#10 we can do this:
```cs
namespace Company.Project
using System;

public class Product
{ 
    ...
}
```

What if this was extended to eliminate another set of curly braces and indentations like this:

```cs
using System;
public class Company.Project.Product;

public void Method1() {} 
/// etc
```

From here the compiler can determine the class to be `Product` and the namespace be `Company.Project`.

**Comment by HaloFour** [comment]
See: https://github.com/dotnet/csharplang/discussions/928

**Reply by CyrusNajmabadi** [reply]
Yup.  And we are *very* unlikely to do this for reasons of intense confusion/ambiguity.  SPecifically, this is not about the language rules (which are likely something we could spec).  This is much about about user confusion as to waht this means and how trivial it is to get bizarre results.

**Reply by KathleenDollard** [reply]
@CyrusNajmabadi 

I'm curious where you think bizarre results would crop up. It seems like just shorthand.

**Reply by jguadagno** [reply]
Me too @CyrusNajmabadi .  I've read the other threads and it just seems like it was dismissed.  I am not suggesting that we eliminate the others but to me, this is similar to Top Level statements but for a class.

**Reply by CyrusNajmabadi** [reply]
@KathleenDollard:


`record class Foo(int x, int y);`

Is that a full definition of a record?  or a file-scoped record?  

If it's the former (which it would have to be because that's how they already shipped), then there's really a lot of pain around:


`class Foo(int x, int y);`

What is this?  A file-scoped class with a primary constructor?  Or a complete class definition that terminates with teh semi-colon (a-la a record).

Records set the standard here that `;` on the type terminates it.  Having other sorts of types (esp. ones that overlap with records) have different semantics there is confusing.

**Reply by jguadagno** [reply]
@CyrusNajmabadi 

In my example, the class namespace is declared in one line, `public class Company.Project.Product;`. I would vision constructors on a new line. Example:

```cs
using System;
public class Company.Project.Product;

public Company.Project.Product(int x, int y) 
// or
public Product(int x, int y

/// rest of class
public void Method1() {} 
/// etc
```

Maybe the syntax only works if you have a namespace in it? What if it is not applicable to record classes?

**Reply by CyrusNajmabadi** [reply]
>  What if it is not applicable to record classes?

That would be an example of a "source of confusion" for me.  It woudl be a very strange thing.  You could have a class like this, but if you added a primary constructor it might mean something else, and the syntax might mean something else entirely when compared to records.

I feel like this is effectively a deal breaker for me.

**Reply by CyrusNajmabadi** [reply]
Also.  wrt:

> public class Company.Project.Product { }

We have another proposal that that be equivalent to:

```c#
partial class Company
{
    partial class Project
    {
        public class Product
        {
        }
    }
}
```

I far prefer this versus thinking that the names are part of a namespace (which is easily written with file-scoped-namespaces).

**Reply by jguadagno** [reply]
I can see this although I, personally, never nest classes.

**Reply by csharpfritz** [reply]
I've read through the other linked issues, and there are more than a few that all are suggesting this concept of file-scoped class defined with the syntax

```
public class MyClass;
```

There is no confusing this with the definition of a record syntax.  Clearly, the record feature uses the keyword `record` and the class feature uses the keyword `class`

The suggestion from @jguadagno and others is that when the keyword `class` is used and the line ends with a semi-colon then the entire contents of the file below that line be wrapped in curly braces.

Single line `record` statements below that `class` statement would be treated as nested records within the declared class.

**Reply by CyrusNajmabadi** [reply]
> There is no confusing this with the definition of a record syntax. 

There is definite confusion.  For example, this is legal today:

```c#
public record class MyClass(int x);
```

Having that be a complete type description if you write that, but now having it be an *open* type description if you write:

```c#
public class MyClass(int x);
```

is extremely *surprising* and non-complementary IMO.

**Reply by csharpfritz** [reply]
I am going to stand by my statement because the legal syntax you are showing includes the keyword `record`.

Some beginners are confused by the positional syntax of `record` and also the block-style syntax of record.  That's a different discussion that has already been resolved as acceptable and now part of the specification.

Why not this?

**Reply by HaloFour** [reply]
There are proposals to expand primary constructors to more than records.

**Reply by csharpfritz** [reply]
Sure... and if I understand the proposed syntax for primary constructors, that syntax includes parenthesis characters `( )` to declare them with appropriate parameters.  Those characters denote a method or a constructor, and in my opinion that's a good standard to keep.

This syntax suggested from the original post does not contain parenthesis characters.

**Reply by HaloFour** [reply]
> and if I understand the proposed syntax for primary constructors, that syntax includes parenthesis characters

Only if the class has a body.  There's nothing about the proposal that requires braces.

**Reply by csharpfritz** [reply]
@HaloFour  Can you help me understand that proposal better by linking it?

**Reply by HaloFour** [reply]
https://github.com/dotnet/csharplang/discussions/4025

**Reply by CyrusNajmabadi** [reply]
> I am going to stand by my statement because the legal syntax you are showing includes the keyword record.

I did cover this in my first response:

> SPecifically, this is not about the language rules (which are likely something we could spec). This is much about about user confusion as to waht this means and how trivial it is to get bizarre results.

This is not a question about if the language can figure this out.  This is a question about how confusing the language would be for consumers.  I have written these language features, and i find this proposal enormously confusing to read and interpret.  Having that single token change the entire meaning of semicolon and how it works is just an anti-pattern for me.  It's a bad thing for the langauge to go that route, regardless of if it could be done or not.

**Reply by CyrusNajmabadi** [reply]
> This syntax suggested from the original post does not contain parenthesis characters.

Right.  But we have to consider how a proposal like this will behave with other proposals we are looking at (and, in this case, are very likely to take).  

Primary constructors have extremely high probability of having.  Easily >90% chance.  So we must consider how this proposal will play along with that.

**Reply by csharpfritz** [reply]
Thinking further about this, you're right about eliminating markup does make the type definition keywords confusing.

What if we pivot slightly and introduce a `file` keyword to indicate that the rest of this file is content of that class?  The syntax originally proposed by @jguadagno could be re-interpreted as:

```
using System;
public file class Company.Project.Product;

public void Method1() {} 
/// etc
```

This could be extended to also provide for 
- `public file struct MyStruct;` 
- `public file enum MyEnum;`

@jguadagno - does that help clarify the intent of your declaration while still allowing the file to drop a pair of braces and the practice of indentation?

**Reply by jguadagno** [reply]
@csharpfritz  I think that could work.

**Comment by KathleenDollard** [comment]
Thanks @CyrusNajmabadi. 

So, we could make the shortcut on files scoped namespaces because we had an explicit new syntax 

    namespace name;  // semi-colon not previously allowed

And we can't do this with classes because the similar approach of using a semi-colon already has meaning on class/struct/record.

Is it reasonable to consider allowing the collapsed namespace./class combination of 

    class namespace.classname
    {...}

**Reply by HaloFour** [reply]
IMO it's more clear with the namespace on its own clause.  The dotted name could be confusing between a namespace and an enclosing class.  Excusing that, it also doesn't really save anything beyond the word "namespace".  At least with file-scope namespaces you were also saving the expected indentation throughout the rest of the file.

**Reply by jnm2** [reply]
@KathleenDollard I would like to work on a proposal where `public class X.Y { }` is shorthand for:

```cs
partial class X
{
    public class Y { }
}
```

In Discord, LDM members were remarking to the same effect as @HaloFour's comment; allowing namespace segments to appear after `class` has pitfalls.

**Comment by orthoxerox** [comment]
A colon could work to disambiguate between a class with no body and a file-scoped class, but this wouldn't match with file-scoped namespaces and would likely be extremely confusing, unless file-scoped classes couldn't have nested classes:

```c#
namespace Acme.Model;

record Foo(int Qux): //Everything below is my members!

record Bar(int Baz); //I am Acme.Model.Foo.Bar
```

```c#
namespace Acme.Model;

record Foo(int Qux);

record Bar(int Baz); //I am Acme.Model.Bar
```

**Reply by FaustVX** [reply]
A colon is really too similar to a semi-colon.
I think file-scoped class shouldn't allow nested class, like file-scoped namespace don't allow nested namespace.

**Reply by CyrusNajmabadi** [reply]
A colon literally cannot work here as `:` already has meaning in that exact position as the base-type-and-interfaces-list separator :)

**Reply by orthoxerox** [reply]
Shoot, I should think harder next time.

**Comment by ryzngard** [comment]
While we're at it, why even include a class name at all? We already recommend them match the file name. Language could allow this to be implicitly derived from the file name. 

```csharp
// File name MyClass.cs
namespace MyNamespace;
public class;

public int I { get; set; }
private void Method1() {}
```

Could be equivalent to 
```csharp
namespace MyNamespace;

public class MyClass
{
    public int I { get; set; }
    private void Method1() {}
}
```

**Reply by jguadagno** [reply]
That could also solve the `record` and constructor issue highlighted earlier
